src/analysis/simulation.jl

```diff
@@ -44,41 +44,41 @@ function simulate(data::Dict{String,Any}, solution::Dict{String,Any},
 
-        if data["node"][node_to_index]["source_id"][2] in tank_set
-            tank_name = data["node"][node_to_index]["source_id"][2]
+        if wm_data["nw"]["1"]["node"][node_to_index]["source_id"][2] in tank_set
+            tank_name = wm_data["nw"]["1"]["node"][node_to_index]["source_id"][2]
             arti_link_dict[tank_name] = artificial_link
-            if data["node"][node_to_index] == tank_index_dict[tank_name]
-                artificial_node = data["node"][node_fr_index]["name"]
+            if wm_data["nw"]["1"]["node"][node_to_index] == tank_index_dict[tank_name]
+                artificial_node = wm_data["nw"]["1"]["node"][node_fr_index]["name"]
             else
-                artificial_node = data["node"][node_to_index]["name"]
+                artificial_node = wm_data["nw"]["1"]["node"][node_to_index]["name"]
             end
             arti_node_dict[tank_name] = artificial_node
-        elseif data["node"][node_fr_index]["source_id"][2] in tank_set
-            tank_name = data["node"][node_fr_index]["source_id"][2]
+        elseif wm_data["nw"]["1"]["node"][node_fr_index]["source_id"][2] in tank_set
+            tank_name = wm_data["nw"]["1"]["node"][node_fr_index]["source_id"][2]
             arti_link_dict[tank_name] = artificial_link
-            if data["node"][node_fr_index] == tank_index_dict[tank_name]
-                artificial_node = data["node"][node_to_index]["name"]
+            if wm_data["nw"]["1"]["node"][node_fr_index] == tank_index_dict[tank_name]
+                artificial_node = wm_data["nw"]["1"]["node"][node_to_index]["name"]
             else
-                artificial_node = data["node"][node_fr_index]["name"]
+                artificial_node = wm_data["nw"]["1"]["node"][node_fr_index]["name"]
             end
             arti_node_dict[tank_name] = artificial_node
         end
     end
 
     link_types = ["pipe", "valve", "pump"]
     for ltype in link_types
-        for (key, link) in solution["1"][ltype]
-            link_name = data[ltype][key]["name"]
-            node_to_index = string(data[ltype][key]["node_to"])
-            node_fr_index = string(data[ltype][key]["node_fr"])
+        for (key, link) in wm_solution["solution"]["nw"]["1"][ltype]
+            link_name = wm_data["nw"]["1"][ltype][key]["name"]
+            node_to_index = string(wm_data["nw"]["1"][ltype][key]["node_to"])
+            node_fr_index = string(wm_data["nw"]["1"][ltype][key]["node_fr"])
 
-            node_to_name = data["node"][node_to_index]["source_id"][2]
-            node_fr_name = data["node"][node_fr_index]["source_id"][2]
+            node_to_name = wm_data["nw"]["1"]["node"][node_to_index]["source_id"][2]
+            node_fr_name = wm_data["nw"]["1"]["node"][node_fr_index]["source_id"][2]
 
             for tank_name in tank_set
                 if ((node_to_name == tank_name) & !(node_fr_name == tank_name)) | (!(node_to_name == tank_name) & (node_fr_name == tank_name))
```

Comment:
Can we document what's happening here? Are you saying that if the link is not an artificial link, it should be added to `tank_link_dict`? Why is this necessary?

Reply:
@tasseff  Here, artificial links refer to those zero-length pipes with shutoff valves added to tanks, and links stored in tank_link_dict are those links that are connected to the tank in the absence of artificial links. I think this part of the codes can be simplified - it looks messy now as I modified it two weeks ago when the artificial node and tank node were reversed in WaterModels. I will try to simplify the codes here and write down some notes next to it.

Reply:
@tasseff , are Yang's revisions sufficient to satisfy you?

Reply:
The top comment that was added helps, thanks.